File: src/models/Emergency.ts
import { CancellationReason } from "./CancellationReason";
import { MissionStatus } from "./MissionStatus";
import { ResponseRating } from "./ResponseRating";
import Team from "./Team";
import { ThreatLevel } from "./ThreatLevel";
import WritableDbItem from "./WritableDbItem";

export default interface Emergency extends WritableDbItem {
  system: string;
  locationId: string | null;
  threatLevel: ThreatLevel;
  clientRsiHandle: string | null;
  clientDiscordId: string | null;
  clientId: string | null;
  status: MissionStatus;
  cancellationReason: CancellationReason;
  refusalReason: string | null;
  alertMessage: DiscordMessage | null;
  clientMessage: DiscordMessage | null;
  coordinationThread: DiscordMessage | null;
  afterActionReportMessage: DiscordMessage | null;
  respondingTeam: Team;
  respondingTeams: RespondingTeam[];
  acceptedOn: string | null;
  completedOn: string | null;
  rating: ResponseRating;
  ratingRemarks: string | null;
  origin: Origin;
  clientData: ClientData | null;
  missionName: string | null;
  isComplete: boolean;
  afterActionReport: AfterActionReport | null;
}

/** A cached Discord message reference. */
export interface DiscordMessage {
  id: string;
  channelId: string;
}

export interface ClientData {
  rsiHandle: string;
  rsiProfileLink: string;
  gotClientData: boolean;
  reported: boolean;
  userSid: string | null;
}

export interface AfterActionReport {
  servicesProvided: MissionServices;
  suspectedTrap: boolean;
  remarks: string | null;
  submitterStaffId: string;
  submittedOn: string | null;
  editHistory: AfterActionReportEdit[];
  hasBeenEdited: boolean;
  resources: AfterActionResource[];
}

export interface AfterActionReportEdit {
  editorStaffId: string;
  editTime: string;
}

export interface AfterActionResource {
  id: string;
  submitterStaffId: string;
  type: AfterActionResourceType;
  url: string;
}

export enum AfterActionResourceType {
  UNKNOWN = 0,
  VIDEO = 1,
}

export enum MissionServices {
  NONE = 0,
  PVE = 1 << 0,
  PVP = 1 << 1,
  REVIVED_HEALED = 1 << 2,
  HEALED_IN_SHIP = 1 << 3,
  EXTRACT_SAFE_ZONE = 1 << 4,
}

export enum Origin {
  UNKNOWN,
  REPORT,
  BEACON,
  EVALUATION,
}

export enum SubmissionSource {
  UNKNOWN,
  API,
  BOT,
}

export interface RespondingTeam {
  id: string;
  teamName: string;
}

File: src-dotnet/models/Emergency.cs
using System;
using System.Collections.Generic;

namespace MedrunnerApiClient.Models
{
    /// <summary>
    /// Represents an emergency mission in the Medrunner system.
    /// </summary>
    public class Emergency : WritableDbItem
    {
        public string System { get; set; }
        public string Subsystem { get; set; }
        public string? TertiaryLocation { get; set; }
        public ThreatLevel ThreatLevel { get; set; }
        public string ClientRsiHandle { get; set; }
        public string? ClientDiscordId { get; set; }
        public string? ClientId { get; set; }
        public string SubscriptionTier { get; set; }
        public MissionStatus Status { get; set; }
        public MessageCache? AlertMessage { get; set; }
        public MessageCache? ClientMessage { get; set; }
        public MessageCache? CoordinationThread { get; set; }
        public MessageCache? AfterActionReportMessage { get; set; }
        public Team RespondingTeam { get; set; }
        public List<RespondingTeam> RespondingTeams { get; set; }
        public long CreationTimestamp { get; set; }
        public long? AcceptedTimestamp { get; set; }
        public long? CompletionTimestamp { get; set; }
        public ResponseRating Rating { get; set; }
        public string? RatingRemarks { get; set; }
        public bool Test { get; set; }
        public CancellationReason CancellationReason { get; set; }
        public string? RefusalReason { get; set; }
        public Origin Origin { get; set; }
        public ClientData? ClientData { get; set; }
        public bool IsComplete { get; set; }
        public string? MissionName { get; set; }
        public AfterActionReport? AfterActionReport { get; set; }
        public SubmissionSource SubmissionSource { get; set; }
    }

    /// <summary>
    /// Represents a cached message reference.
    /// </summary>
    public class MessageCache
    {
        public string Id { get; set; }
        public string ChannelId { get; set; }
    }

    /// <summary>
    /// Represents client data for an emergency.
    /// </summary>
    public class ClientData
    {
        public string RsiHandle { get; set; }
        public string RsiProfileLink { get; set; }
        public bool GotClientData { get; set; }
        public bool RedactedOrgOnProfile { get; set; }
        public bool Reported { get; set; }
    }
}
